src/server.rs

```diff
@@ -1,31 +1,35 @@
+use std::sync::atomic::{AtomicUsize, Ordering};
+
 use anyhow::{Context, Result};
-use tokio::net::TcpListener;
 use tokio::task;
 use tracing::{error, info, info_span, warn, Instrument};
 
 use crate::client;
 use crate::config::Config;
 use crate::instance::InstanceMap;
+use crate::socketwrapper::Listener;
 
 pub async fn run(config: &Config) -> Result<()> {
     let instance_map = InstanceMap::new(config).await;
+    let next_client_id = AtomicUsize::new(0);
 
-    let listener = TcpListener::bind(config.listen).await.context("listen")?;
+    let listener = Listener::bind(&config.listen).await
+    .context("listen")?;
     loop {
         match listener.accept().await {
-            Ok((socket, addr)) => {
-                let port = addr.port();
+            Ok((socket, _addr)) => {
+                let client_id = next_client_id.fetch_add(1, Ordering::Relaxed);
```

Comment:
```suggestion
                let client_id = next_client_id();
```

i'd write it like this (together with the suggestion above). nothing is wrong with your code, but it's my style preference to express that the `next_client_id` atomic isn't used elsewhere by shadowing it immediately.